Check etcd quorum for Control Plane Remediation/Maintenance

razo7 opened the pull request:
New function (with tests) for ETCD quorum violation check.
It can accept/refuse Control Plane remediation/maintenance (NHC/NMO respectively) for any control plane node.
It begins with checking if etcd quorum exists and it allows disruptions.
If it exists but no disruptions are allowed then it checks if the given node is disrupted (guard pod status is _False_). When the given node is indeed disrupted and no etcd quorum violations are accepted, then the node can be disrupted as it won't violate the quorum, otherwise, any other control plane node disruption will violate the quorum.

razo7 committed:
Add IsEtcdDisruptionAllowed to enable/disable control plane remediation or maintenance

Will be used by NMO and NHC to verify whether ETCD quorum was violated and if the desired node is the one who made this violation

razo7 committed:
Add test for IsEtcdDisruptionAllowed to check healthy cluster and one unhealthy control plane node

razo7 committed:
Split one function into two for better code separation

IsControlPlaneNodeIsReady reflects better what the function does while only fraction of the function is for checking etcd pdb disruptions

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// IsControlPlaneNodeReady checks if etcd disruption is allowed and accpet/refuse the todoAction (remediation/manitenance)
+func IsControlPlaneNodeReady(ctx context.Context, cl client.Client, node *corev1.Node, todoAction string) (bool, error) {
```
- function name is misleading, we do not check if the node is ready. We just check if we can disrupt the given node without disrupting etcd quorum
- while usually it makes sense to split longer functions into more smaller ones, here I don't see any value. `isEtcdDisruptionAllowed` needing to return the pdb itself is a bit strange at least
- passing the `todoAction` is bad separation of concerns. Better limit logs to what we do and know here, and let the caller log the consequences.
- `etcd` should be lower case in logs, it's a project name and not an abbreviation for something.

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,81 @@
+	return false, &pdb, nil
+}
+
+// IsControlPlaneNodeReady checks if etcd disruption is allowed and accpet/refuse the todoAction (remediation/manitenance)
```
please add a comment about which permissions this needs

```
// +kubebuilder:rbac:groups=policy,resources=poddisruptionbudgets,verbs=get;list;watch
// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
```

razo7 replied:
But if we add it here will the other project that calls for the `IsControlPlaneNodeReady` function will use these Kubebuilder lines to update RBAC?
Should these lines be written locally in the parent project or is it needed also here?

slintes replied:
I don't think kubebuilder looks at vendored code, so it won't have any effect. What I mean is that we should document that the parent project needs to add these lines

razo7 replied:
>I don't think kubebuilder looks at vendored code, so it won't have any effect.

+1

> What I mean is that we should document that the parent project needs to add these lines

SGTM

razo7 committed:
Check whether a node can disrupt etcd

Better one function name, typos, improve logs, and be more generic

razo7 committed:
Comment about adding kubebuilder permissions in parent project

These two lines will be discarded from modifing RBAC, thus a user must manually add them to his project, so CanNodeDisruptEtcd can properly run

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,79 @@
+var log = logf.Log.WithName("etcd-pdb-checker")
+
+// CanNodeDisruptEtcd checks if a node can disrupt etcd quorum, and it returns error when it fails in the process
+func CanNodeDisruptEtcd(ctx context.Context, cl client.Client, node *corev1.Node) (bool, error) {
```
Can node disrupt etcd?
The return values report the opposite 🤷🏼 E.g. when DisruptionsAllowed >= 1 you can and will not disrupt the etcd quorum...

Why moving away from the isSomethingAllowed name? It's clear and sets expectation about the meaning of the return value right, not?

razo7 replied:
>Why moving away from the isSomethingAllowed name? It's clear and sets expectation about the meaning of the return value right, not?

I haven't found the right word for `isSomethingAllowed` format :( Function naming is sometimes harder than the function logic ;)

slintes replied:
When you don't have an idea for a better name, maybe just keep what we have in NHC? 🤷🏼‍♂️

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,79 @@
+		return false, err
+	}
+	if len(pdbList.Items) == 0 {
+		log.Info("No PDB were found, "+errNoEtcdChec, "namespace", etcdNamespace)
```
- when you add `were` to `No PDB were found` you switch from singular to plural, so it should be `PDBs` 🤷🏼‍♂️ 
- please don't use string concatenation with `+`, we have `something*f*()` methods for that...

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,79 @@
+	errNoEtcdChec = "can't check if etcd quorum will be violated!"
+)
+
+var log = logf.Log.WithName("etcd-pdb-checker")
```
We typically use `github.com/go-logr/logr.Logger` in our code. It's probably best t pass a logger as argument to the function call, that way we keep already set information and get better logs.

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,79 @@
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Important: add the following two lines in your project/code, so your RBAC will be updated with the right permissions
```
for better visibility please make this a package or function comment

razo7 committed:
Remove redundant comments

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,79 @@
+					return true, nil
+				}
+			}
+			log.Info("Node is not disrupted, but it will violate etcd quorum", "Node", node.Name, "Guard pod", pod.Name)
```
s/but/and

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,79 @@
+		}
+	}
+
+	log.Info("Node is not disrupted, but it will violate etcd quorum", "Node", node.Name)
```
when there is no pod for the given node, it is either already disrupted, or it isn't configured to run etcd, so the quorum will not be violated

razo7 replied:
> it is either already disrupted

Do you mean that when a control plane node is disrupted then its guard pod could be missing?

slintes replied:
> when a control plane node is disrupted then its guard pod could be missing?

yes, when remediation rebooted the node and the pod didn't restart yet

slintes commented on pkg/etcd/etcd_suite_test.go:
```diff
@@ -0,0 +1,86 @@
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+// var (
```
Please remove commented code blocks. No one will keep it up to date, it's better to recreate or copy from another place when needed

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,163 @@
+		cl client.WithWatch
+	)
+
+	When("etcd PDB was not set", func() {
```
what do you mean with "was not set"?

razo7 replied:
was not created or when this PDB is missing.

slintes replied:
maybe "etcd PDB doesn't exist" would be better then?

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,163 @@
+})
+
+// getPogGuard returns gurad pod with expected label and Ready condition is True for a given nodeName
+func getPogGuard(nodeName string) *corev1.Pod {
```
s/Pog/Pod ?

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,163 @@
+
+func newControlPlaneNodes(nodesCount int) []*corev1.Node {
+	nodes := make([]*corev1.Node, 0, nodesCount)
+	for i := nodesCount; i > 0; i-- {
```
why reverse order? 🤔 🙂 
usually you use `for i := 0; i < count; i++`...

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,163 @@
+			Labels: labels,
+		},
+		Status: corev1.NodeStatus{
+			Conditions: []corev1.NodeCondition{
```
for what do we need node conditions? 🤔

clobrano commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,79 @@
+
+const (
+	etcdNamespace = "openshift-etcd"
+	errNoEtcdChec = "can't check if etcd quorum will be violated!"
```
typo, I guess, it should be errNoEtcdCheck

razo7 committed:
When guard pod is missing then we won't violate the etcd quorum

Use log as input, fmt.Sprintf instead of concating strings, and better function names, logs, typos

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,176 @@
+	var (
+		cl client.WithWatch
+	)
+	zapLogger, _ := zap.NewProduction()
```
should be NewDevelopment in order to print more logs

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,176 @@
+		cl client.WithWatch
+	)
+	zapLogger, _ := zap.NewProduction()
+	log := zapr.NewLogger(zapLogger)
```
we need to write to GinkWriter, and better time formatting would be great as well :) See e.g. https://github.com/medik8s/node-healthcheck-operator/blob/main/controllers/suite_test.go#L104-L108

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,176 @@
+			nodeGuardDown = controlPlaneNodes[2].Name
+			pdb.Status.DisruptionsAllowed = int32(0)
+			Expect(cl.Status().Update(context.Background(), pdb)).To(Succeed())
+			dummyPodGuard := getPodGuard(nodeGuardDown)
```
this isn't needed, you can do the same directly with podGuard

slintes reviewed (changes requested):
Few nits left.

Test could be structured better, e.g. modifiying DisruptionAllowed in It(), while it was set in BeforeEach() isn't that nice. A new (maybe nested) Context / When with its own BeforeEach would be a bit nicer. But I won't block on it, it's ok like it is.

razo7 committed:
Reconstruct tests and creating namespace

Add context and more when clauses for better logic. Clarify log messages, and create etcd namespace for each test

slintes commented on pkg/etcd/etcd.go:
```diff
@@ -0,0 +1,80 @@
+// etcd package is used for checking whether etcd disruption is allowd
```
package comments should start with "Package <package name>"
https://tip.golang.org/doc/comment

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,198 @@
+		Development: true,
+		TimeEncoder: zapcore.RFC3339NanoTimeEncoder,
+	}
+	log := zaplog.New(zaplog.WriteTo(GinkgoWriter), zaplog.UseFlagOptions(&opts))
```
This should go into BeforeSuite.
Actually, we already have something similar over there...?

razo7 replied:
Are you referring to the log or the fake controller?

slintes replied:
log

razo7 replied:
To create a log per test? Isn't one log would be enough for all the unit tests?

slintes replied:
one log is enough, yes, and that's why it can or even should go into BeforeSuite, where we already have some log setup, which is duplicated here

slintes commented on pkg/etcd/etcd_test.go:
```diff
@@ -0,0 +1,198 @@
+			})
+		})
+
+		When("etcd PDB allowed disruptions is one with a gurad pod per control plane node", func() {
```
typo gurad

razo7 committed:
Use Package name and refactor fakeClient

Package comment should begin with the word 'Package', initalize fakeClient at BeforeEach, and typo

razo7 committed:
Setup log in BeforeSuite

One log creation is enough, and it is better to create it at BeforeSuite
